### PR TITLE
Build an AppImage for each git push

### LINKED ISSUE #3056: Provide an AppImage for many Linux distributions
Providing an [AppImage](http://appimage.org/) would have, among others, these advantages:
- Works for most Linux distributions (including Ubuntu, Fedora, openSUSE, CentOS, elementaryOS, Linux Mint, and others)
- One app = one file = super simple for users: just download one AppImage file, [make it executable](http://discourse.appimage.org/t/how-to-make-an-appimage-executable/80), and run
- No unpacking or installation necessary
- No root needed
- No system libraries changed
- Works out of the box, no installation of runtimes needed
- Optional desktop integration with `appimaged`
- Optional binary delta updates, e.g., for continuous builds (only download the binary diff) using AppImageUpdate
- Can optionally GPG2-sign your AppImages (inside the file)
- Works on Live ISOs
- Can use the same AppImages when dual-booting multiple distributions

[Here is an overview](https://github.com/probonopd/AppImageKit/wiki/AppImages) of projects that are already distributing upstream-provided, official AppImages.

I have produced an experimental, proof-of-concept AppImage by converting the binaries from the Ubuntu trusty PPA to AppImage using this file
https://github.com/probonopd/AppImages/blob/master/recipes/meta/FontForge.yml

It can be downloaded from
https://bintray.com/probono/AppImages/FontForge#files

If you have questions, AppImage developers are on #AppImage on irc.freenode.net.

### PR DESCRIPTION
**Pull request by probonopd**
Screenshot: This is the FontForge AppImage built on Travis CI from `master`, running on Clear Linux (a distribution for which no native FontForge package exists)

### Motivation and Context
- [ ] Why is this change required? What problem does it solve?
- [x] Closes #3056

### Types of changes
What types of changes does your code introduce? Check all the boxes that apply:
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Description
- [x] Describe your changes in detail.

This PR, when merged, will use the [Travis CI](https://travis-ci.org/) build to create an [AppImage](http://appimage.org/) for each `git push`.

The AppImage could then be uploaded to GitHub Releases automatically.

Providing an [AppImage](http://appimage.org/) would have, among others, these advantages:

- Applications packaged as an AppImage can run on many distributions (including Ubuntu, Fedora, openSUSE, CentOS, elementaryOS, Linux Mint, and others)
- One app = one file = super simple for users: just download one AppImage file, [make it executable](http://discourse.appimage.org/t/how-to-make-an-appimage-executable/80), and run
- No unpacking or installation necessary
- No root needed
- No system libraries changed
- Works out of the box, no installation of runtimes needed
- Optional desktop integration with `appimaged`
- Optional binary delta updates, e.g., for continuous builds (only download the binary diff) using AppImageUpdate
- Can optionally GPG2-sign your AppImages (inside the file)
- Works on Live ISOs
- Can use the same AppImages when dual-booting multiple distributions
- Can be listed in the [AppImageHub](https://appimage.github.io/apps) central directory of available AppImages
- Can double as a self-extracting compressed archive with the `--appimage-extract` parameter
- No repositories needed. Suitable/optimized for air-gapped (offline) machines

[Here is an overview](https://appimage.github.io/apps) of projects that are already distributing upstream-provided, official AppImages.

If you have questions, AppImage developers are on #AppImage on irc.freenode.net.

### Final checklist

- [x] My code follows the code style of this project found [here](https://github.com/fontforge/fontforge/blob/master/CONTRIBUTING.md#coding-style).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/fontforge/fontforge/blob/master/CONTRIBUTING.md) guidelines.

### CODE REVIEW

**Comment by frank-trampe** [comment]
@JoesCat, can you review this for compatibility with your recent changes to the theming?

**Comment by JoesCat** [comment]
On initial preview, I would prefer to see this as another branch, example, no-gui, full-linux, appbuild, osx. This way, we still get the other tests to run.
I'll dowload this, but not sure about getting this modified and merged before year-end.

**Comment by probonopd** [comment]
I think the main point about having continuous AppImages is that testers can test every build of master, and even every pull request before it is merged. If this only lives in some branch, then we would not get these advantages. But why not build the AppImage __and__ run all the other tests _in addition_ to AppImage generation in `master`?

File: .travis.yml [hunk 1]
```diff
@@ -122,13 +120,39 @@ install:
...
-  - make install
+  - make -j$(nproc)
+  - if [ "$TRAVIS_OS_NAME" == "linux" ]; then mkdir -p appdir ; make install DESTDIR=$(readlink -f appdir); fi
+  - if [ "$TRAVIS_OS_NAME" != "linux" ]; then make install; fi
   - make check
   - if [ ! -z "$LINUX_FULL" ]; then make distcheck; fi
   - fontforge -version
   - $PYTHON -c "import fontforge; import psMat; print(fontforge.__version__, fontforge.version()); fontforge.font();"
   - if [ "$TRAVIS_OS_NAME" == "osx" ] && [ "$TRAVIS_PULL_REQUEST" == "false" ] && [ "$TRAVIS_BRANCH" == "master" ]; then travis-scripts/ffosxbuild.sh $PREFIX $TRAVIS_COMMIT; fi
+  - |
```
Comment (jtanx) [inline .travis.yml]: Move all of this into a packaging script, similar to `travis-scripts/ffosxbuild.sh`

File: .travis.yml [hunk 2]
```diff
@@ -122,13 +120,39 @@ install:
...
+    chmod a+x linuxdeployqt-continuous-x86_64.AppImage
+    export VERSION=$(git rev-parse --short HEAD) # linuxdeployqt uses this for naming the file
+    ./linuxdeployqt*.AppImage ./appdir/usr/share/applications/*.desktop -bundle-non-qt-libs
+    # Manually invoke appimagetool so that the custom AppRun stays intact
+    ./linuxdeployqt*.AppImage --appimage-extract
+    export PATH=$(readlink -f ./squashfs-root/usr/bin):$PATH
+    rm appdir/AppRun ; cp Packaging/AppDir/AppRun appdir/AppRun ; chmod +x appdir/AppRun # custom AppRun
+    ./squashfs-root/usr/bin/appimagetool -g ./appdir/
+    find appdir -executable -type f -exec ldd {} \; | grep " => /usr" | cut -d " " -f 2-3 | sort | uniq
+    curl --upload-file FontForge*.AppImage https://transfer.sh/FontForge-git.$(git rev-parse --short HEAD)-x86_64.AppImage # For testing only
```
Comment (jtanx) [inline .travis.yml]: You can upload to bintray, we do this for the osx CI builds

File: .travis.yml [hunk 3]
```diff
@@ -54,9 +48,13 @@ install:
   - |
     set -e
     export DEPSPREFIX=$TRAVIS_BUILD_DIR/travisdeps
-    export PREFIX=$TRAVIS_BUILD_DIR/target
-    export PATH=$PATH:$DEPSPREFIX/bin:$PREFIX/bin
-    export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$DEPSPREFIX/lib:$PREFIX/lib
+    if [ "$TRAVIS_OS_NAME" == "linux" ]; then
```
Comment (jtanx) [inline .travis.yml]: The cleaner way to do this is to always set the prefix to `/usr` then set the destdir on `make install` to `$TRAVIS_BUILD_DIR/target`. Your packaging script should then be responsible for transforming from that instead of creating the `appdir` folder  up front

File: .travis.yml [hunk 4]
```diff
@@ -1,15 +1,9 @@
 language: c
 compiler: gcc
-sudo: false
+sudo: true
 dist: trusty
 osx_image: xcode9.2
 
-branches:
```
[the remark this answers is not in the document]
Comment (probonopd) [reply .travis.yml]: I think I removed this so that during development of the PR things got built. If everything else is ready for merging, this can be added back.

File: .travis.yml [hunk 5]
```diff
@@ -1,15 +1,9 @@
 language: c
 compiler: gcc
-sudo: false
+sudo: true
```
[the remark this answers is not in the document]
Comment (probonopd) [reply .travis.yml]: I always do this so that I can `sudo apt-get install` things, but it seems not to be needed here, so removing.

File: .travis.yml (hunk 1 above)
Comment (probonopd) [reply .travis.yml]: Moved there. Personally I find it easier if everything is done directly in `.travis.yml` but it's a matter of personal preference.

File: .travis.yml (hunk 2 above)
Comment (probonopd) [reply .travis.yml]: How do we tell `provider: bintray` which files to upload?

File: .travis.yml (hunk 3 above)
Comment (probonopd) [reply .travis.yml]: Using '$TRAVIS_BUILD_DIR/target' as the AppDir now.

File: .travis.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (jtanx) [reply .travis.yml]: Yes please, add this back before merging. All PRs should still build with this in place, but it just restricts non-PR branch builds to those listed.

File: .travis.yml (hunk 3 above)
Comment (jtanx) [reply .travis.yml]: This wasn't my intent, what I meant was something along the lines of

```sh
export PREFIX=/usr #unconditionally
export TARGET_DIR=$TRAVIS_BUILD_DIR/target
export PATH=$PATH:$DEPSPREFIX/bin:$TARGET_DIR/bin
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$DEPSPREFIX/lib:$TARGET_DIR/lib
```

(effectively replacing PREFIX with TARGET_DIR)

then at make

```
...
make install DESTDIR=$(readlink -f $TARGET_DIR) #unconditionally
``

File: .travis.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (probonopd) [reply .travis.yml]: Thanks, didn't know that it does not affect PRs. Changed.

File: .travis.yml (hunk 3 above)
Comment (probonopd) [reply .travis.yml]: Currently there is

https://github.com/fontforge/fontforge/blob/a05dc286ade51c245f65b591f15dee85dcced172/.travis.yml#L57

I did not want to change that so as to not affect non-Linux builds in any way.

**Comment by frank-trampe** [comment]
@probonopd, just so you know, @JoesCat is having problems with his Internet service provider but will likely be back soon. I still plan to get this merged before the release next week.

File: .travis.yml (hunk 3 above)
Comment (jtanx) [reply .travis.yml]: I'll take a look at this later today

**Comment by JoesCat** [comment]
Hi @probonopd - since this would be considered a user app, not a compiler test, you can leave libunicodenames out of the formula. It has not been updated in years and there are well over a few thousand new characters added since then. Best to update travis to pull the latest libuninameslist as it's at unicode11.0 (From a user point of view, it runs faster too since all references are computed at compile time, not run time, so it may be noticed on some older or slower CPU architectures).

**Comment by probonopd** [comment]
Hi @JoesCat I did not change anything regarding libunicodenames, just left it as it was. Would it be OK for you to make the Unicode changes after we have AppImage generation merged?

**Comment by JoesCat** [comment]
Thank @jtanx - won't be seeing internet properly until Monday.... otherwise got to say it's a quiet holiday ;-)

File: .travis.yml (hunk 3 above)
Comment (jtanx) [reply .travis.yml]: From my experiments, we don't actually need to do the whole prefix/destdir shenanigans, the appimage bundling thingo handles relocating the dlls just fine with the current prefix setup. @probonopd  take a look at the modifications and see what you think

File: .travis.yml [hunk 6]
```diff
@@ -54,13 +54,9 @@ install:
...
     set -e
     export DEPSPREFIX=$TRAVIS_BUILD_DIR/travisdeps
-    if [ "$TRAVIS_OS_NAME" == "linux" ]; then
-      export PREFIX=/usr
-    else
-      export PREFIX=$TRAVIS_BUILD_DIR/target
-    fi
-    export PATH=$PATH:$DEPSPREFIX/bin:$PREFIX/bin:$TRAVIS_BUILD_DIR/target/$DEPSPREFIX/bin:$TRAVIS_BUILD_DIR/target/$PREFIX/bin
-    export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$DEPSPREFIX/lib:$PREFIX/lib:$TRAVIS_BUILD_DIR/target/$DEPSPREFIX/lib:$TRAVIS_BUILD_DIR/target/$PREFIX/lib
+    export PREFIX=$TRAVIS_BUILD_DIR/target
```
[the remark this answers is not in the document]
Comment (jtanx) [reply .travis.yml]: That doesn't seem to be the case for the appimage that was generated, see e.g. https://transfer.sh/C2U5I/FontForge-git.e22d188-x86_64.AppImage

Running ldd on the extracted dlls shows that it's been fixed up to /usr

```
jeremy@pickle:~/fontforge/tmp/squashfs-root/usr/lib$ ldd libfontforgeexe.so
	linux-vdso.so.1 (0x00007ffc36343000)
	libgunicode.so.4 => /usr/local/lib/libgunicode.so.4 (0x00007f17badb1000)
	libgutils.so.2 => /usr/local/lib/libgutils.so.2 (0x00007f17bab89000)
	libgdraw.so.5 => /usr/local/lib/libgdraw.so.5 (0x00007f17ba877000)
	libfreetype.so.6 => /usr/lib/x86_64-linux-gnu/libfreetype.so.6 (0x00007f17ba5c8000)
	libgobject-2.0.so.0 => /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0 (0x00007f17ba375000)
	libglib-2.0.so.0 => /lib/x86_64-linux-gnu/libglib-2.0.so.0 (0x00007f17ba061000)
	libfontforge.so.2 => /usr/local/lib/libfontforge.so.2 (0x00007f17b997f000)
	libzmqcollab.so.2 => /home/jeremy/fontforge/tmp/squashfs-root/usr/lib/./libzmqcollab.so.2 (0x00007f17b976b000)
	libczmq.so.3 => /home/jeremy/fontforge/tmp/squashfs-root/usr/lib/./libczmq.so.3 (0x00007f17b94f2000)
	libzmq.so.5 => /home/jeremy/fontforge/tmp/squashfs-root/usr/lib/./libzmq.so.5 (0x00007f17b9285000)
	libpython3.6m.so.1.0 => /home/jeremy/fontforge/tmp/squashfs-root/usr/lib/./libpython3.6m.so.1.0 (0x00007f17b8bb9000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f17b88b5000)
	libltdl.so.7 => /home/jeremy/fontforge/tmp/squashfs-root/usr/lib/./libltdl.so.7 (0x00007f17b86aa000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f17b848d000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f17b80ee000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f17b7eea000)
	libgif.so.7 => /usr/lib/x86_64-linux-gnu/libgif.so.7 (0x00007f17b7ce0000)
	libjpeg.so.62 => /usr/lib/x86_64-linux-gnu/libjpeg.so.62 (0x00007f17b7a75000)
	libpng16.so.16 => /usr/lib/x86_64-linux-gnu/libpng16.so.16 (0x00007f17b7842000)
	libtiff.so.5 => /usr/lib/x86_64-linux-gnu/libtiff.so.5 (0x00007f17b75cb000)
	libuninameslist.so.1 => /usr/local/lib/libuninameslist.so.1 (0x00007f17b70f3000)
	libpangocairo-1.0.so.0 => /usr/lib/x86_64-linux-gnu/libpangocairo-1.0.so.0 (0x00007f17b6ee6000)
	libpango-1.0.so.0 => /usr/lib/x86_64-linux-gnu/libpango-1.0.so.0 (0x00007f17b6c9a000)
	libcairo.so.2 => /usr/lib/x86_64-linux-gnu/libcairo.so.2 (0x00007f17b6986000)
	libreadline.so.7 => /lib/x86_64-linux-gnu/libreadline.so.7 (0x00007f17b6739000)
	libspiro.so.0 => /usr/local/lib/libspiro.so.0 (0x00007f17b6534000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f17b631a000)
	libgio-2.0.so.0 => /usr/lib/x86_64-linux-gnu/libgio-2.0.so.0 (0x00007f17b5f84000)
	libxml2.so.2 => /usr/lib/x86_64-linux-gnu/libxml2.so.2 (0x00007f17b5bc9000)
	libpangoxft-1.0.so.0 => /usr/lib/x86_64-linux-gnu/libpangoxft-1.0.so.0 (0x00007f17b59c0000)
	libpangoft2-1.0.so.0 => /usr/lib/x86_64-linux-gnu/libpangoft2-1.0.so.0 (0x00007f17b57aa000)
	libfontconfig.so.1 => /usr/lib/x86_64-linux-gnu/libfontconfig.so.1 (0x00007f17b556c000)
	libXft.so.2 => /usr/lib/x86_64-linux-gnu/libXft.so.2 (0x00007f17b5356000)
	libwoff2enc.so.1.0.2 => /usr/lib/x86_64-linux-gnu/libwoff2enc.so.1.0.2 (0x00007f17b5147000)
	libbrotlienc.so.1 => /usr/lib/x86_64-linux-gnu/libbrotlienc.so.1 (0x00007f17bbb64000)
	libwoff2dec.so.1.0.2 => /usr/lib/x86_64-linux-gnu/libwoff2dec.so.1.0.2 (0x00007f17b4f3c000)
	libbrotlidec.so.1 => /usr/lib/x86_64-linux-gnu/libbrotlidec.so.1 (0x00007f17bbb53000)
	libSM.so.6 => /usr/lib/x86_64-linux-gnu/libSM.so.6 (0x00007f17b4d34000)
	libICE.so.6 => /usr/lib/x86_64-linux-gnu/libICE.so.6 (0x00007f17b4b17000)
	libX11.so.6 => /usr/lib/x86_64-linux-gnu/libX11.so.6 (0x00007f17b47d7000)
	libffi.so.6 => /usr/lib/x86_64-linux-gnu/libffi.so.6 (0x00007f17b45ce000)
	libpcre.so.3 => /lib/x86_64-linux-gnu/libpcre.so.3 (0x00007f17b435b000)
	libgioftp.so.2 => /usr/local/lib/libgioftp.so.2 (0x00007f17b4155000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f17b3f4d000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f17b3bcb000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f17b39b4000)
	libexpat.so.1 => /lib/x86_64-linux-gnu/libexpat.so.1 (0x00007f17b378a000)
	libutil.so.1 => /lib/x86_64-linux-gnu/libutil.so.1 (0x00007f17b3587000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f17bb9f4000)
	liblzma.so.5 => /lib/x86_64-linux-gnu/liblzma.so.5 (0x00007f17b3361000)
	libjbig.so.0 => /usr/lib/x86_64-linux-gnu/libjbig.so.0 (0x00007f17b3153000)
	libgthread-2.0.so.0 => /usr/lib/x86_64-linux-gnu/libgthread-2.0.so.0 (0x00007f17b2f51000)
	libthai.so.0 => /usr/lib/x86_64-linux-gnu/libthai.so.0 (0x00007f17b2d47000)
	libpixman-1.so.0 => /usr/lib/x86_64-linux-gnu/libpixman-1.so.0 (0x00007f17b2aa0000)
	libxcb-shm.so.0 => /usr/lib/x86_64-linux-gnu/libxcb-shm.so.0 (0x00007f17b289c000)
	libxcb.so.1 => /usr/lib/x86_64-linux-gnu/libxcb.so.1 (0x00007f17b2674000)
	libxcb-render.so.0 => /usr/lib/x86_64-linux-gnu/libxcb-render.so.0 (0x00007f17b2466000)
	libXrender.so.1 => /usr/lib/x86_64-linux-gnu/libXrender.so.1 (0x00007f17b225c000)
	libXext.so.6 => /usr/lib/x86_64-linux-gnu/libXext.so.6 (0x00007f17b204a000)
	libtinfo.so.5 => /lib/x86_64-linux-gnu/libtinfo.so.5 (0x00007f17b1e20000)
	libgmodule-2.0.so.0 => /usr/lib/x86_64-linux-gnu/libgmodule-2.0.so.0 (0x00007f17b1c1c000)
	libselinux.so.1 => /lib/x86_64-linux-gnu/libselinux.so.1 (0x00007f17b19f4000)
	libresolv.so.2 => /lib/x86_64-linux-gnu/libresolv.so.2 (0x00007f17b17dd000)
	libmount.so.1 => /lib/x86_64-linux-gnu/libmount.so.1 (0x00007f17b158f000)
	libicui18n.so.57 => /usr/lib/x86_64-linux-gnu/libicui18n.so.57 (0x00007f17b1115000)
	libicuuc.so.57 => /usr/lib/x86_64-linux-gnu/libicuuc.so.57 (0x00007f17b0d6d000)
	libicudata.so.57 => /usr/lib/x86_64-linux-gnu/libicudata.so.57 (0x00007f17af2f0000)
	libharfbuzz.so.0 => /usr/lib/x86_64-linux-gnu/libharfbuzz.so.0 (0x00007f17af05b000)
	libwoff2common.so.1.0.2 => /usr/lib/x86_64-linux-gnu/libwoff2common.so.1.0.2 (0x00007f17aee58000)
	libbrotlicommon.so.1 => /usr/lib/x86_64-linux-gnu/libbrotlicommon.so.1 (0x00007f17bbb24000)
	libuuid.so.1 => /lib/x86_64-linux-gnu/libuuid.so.1 (0x00007f17aec53000)
	libbsd.so.0 => /lib/x86_64-linux-gnu/libbsd.so.0 (0x00007f17aea3d000)
	libdatrie.so.1 => /usr/lib/x86_64-linux-gnu/libdatrie.so.1 (0x00007f17ae835000)
	libXau.so.6 => /usr/lib/x86_64-linux-gnu/libXau.so.6 (0x00007f17ae631000)
	libXdmcp.so.6 => /usr/lib/x86_64-linux-gnu/libXdmcp.so.6 (0x00007f17ae42b000)
	libblkid.so.1 => /lib/x86_64-linux-gnu/libblkid.so.1 (0x00007f17ae1e5000)
	libgraphite2.so.3 => /usr/lib/x86_64-linux-gnu/libgraphite2.so.3 (0x00007f17adfb8000)

```

File: .travis.yml [hunk 7]
```diff
@@ -127,14 +123,13 @@ script:
   - ./bootstrap
   - ./configure $FFCONFIG
   - make -j$(nproc)
-  - if [ "$TRAVIS_OS_NAME" == "linux" ]; then mkdir -p $TRAVIS_BUILD_DIR/target ; make install DESTDIR=$(readlink -f $TRAVIS_BUILD_DIR/target); fi
-  - if [ "$TRAVIS_OS_NAME" != "linux" ]; then make install; fi
+  - make install
```
[the remark this answers is not in the document]
Comment (jtanx) [reply .travis.yml]: Sure, but it's easier without doing that, and having a prefix but installing to a different destdir just doesn't work on mac

File: .travis.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (probonopd) [reply .travis.yml]:
```
me@host:~$ grep -r /home/travis/build/fontforge /tmp/.mount_FontFoCwNvLZ | wc -l
53
```

tells me that the string `/home/travis/build/fontforge` is hardcoded in at least 53 places.

[the remark this answers is not in the document]
Comment (jtanx) [reply .travis.yml]: I think I'll live with that, it still runs just fine.

File: .travis.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (probonopd) [reply .travis.yml]: Which is why I was limiting that to Linux

File: .travis.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (probonopd) [reply .travis.yml]: OK, if it still works...

File: desktop/Makefile.am [hunk 8]
```diff
@@ -39,6 +39,8 @@ dist_hicolor256x256_DATA = icons/256x256/apps/fontforge.png
 dist_hicolorscalable_DATA = icons/scalable/apps/fontforge.svg
 dist_mimepackages_DATA = fontforge.xml
 
+if XDG_DESKTOP
+
 install-data-hook: update-mime-database update-desktop-database
 uninstall-hook: update-mime-database update-desktop-database
```
Comment (khaledhosny) [inline desktop/Makefile.am]: These two rules should drop this too, it isn’t the job of Makefile to call these tools and I don’t know any other package that does this.

Comment (jtanx) [reply desktop/Makefile.am]: I can see a few projects on github do this, but then again I'm not too familiar with this. If it's not done here, is it just run automatically or something?

Comment (JoesCat) [reply desktop/Makefile.am]: Yes, there are a couple examples, this is a solid example using it:
https://github.com/dleidert/chemical-mime/blob/master/freedesktop.org/Makefile.am

This shows puzzlement with it, and apparently distro sensitive (Ubuntu? therefore Debian too):
https://github.com/byteball/obyte-gui-wallet/issues/17
These utilities are default installed on Mageia, and therefore I suspect may also be within Redhat, Fedora, SuSE, Centos...

For example, GIMP, will validate (dist), and only touch the directory:
https://github.com/GNOME/gimp/blob/master/desktop/Makefile.am

...and there is good documentation going into this sort of stuff (circa 2008..2014 -ish timeline):
https://help.ubuntu.com/community/AddingMimeTypes

This seems like a good explanation:
https://askubuntu.com/questions/18806/why-does-update-mime-database-complain-about-uri-rtspt-and-other-unusual-types
"That's a problem with update-mime-database and the MIME standard itself. Originally it...."

...so if you're confused about what to do about all this, you're not alone  ;-P
Reading git blame, Abel Cheung was quite capable and saw good reason to include it.
...and just because some distros don't make use of these, means we should make exception to allow it to quietly pass along, but at the same time, make use of it for those that have it.
In Summary, if it's there, use it, if not there, skip it.

...more reading shows @probonopd has gone through a fair bit of this, so I would lean in the direction of what's already in this PR
https://github.com/AppImage/appimaged/issues/34

Comment (khaledhosny) [reply desktop/Makefile.am]: They should be run automatically by the desktop environment. Calling these functions modifies files on the system that can’t be undone, that is one of the `make disthcheck` failures that I have seen last time I tried it (a good autotools project should always pass `make distcheck`).

**Comment by JoesCat** [comment]
@frank-trampe:
> @JoesCat, can you review this for compatibility with your recent changes to the theming?

Best to merge a working version of this first, then, apply an updated version of the other patch over so we get some of the benefits as well. At the time I started the other PR, it was when it had an older travis file. My PR can be a template to a revised improvement....but this PR needs our attention first.

File: desktop/Makefile.am (hunk 8 above)
Comment (JoesCat) [reply desktop/Makefile.am]: Unfortunately, we still have not reached a stable make distcheck yet. If I ran 'make dist' and then unzip the dist tar into a new directory, there are still a lot of missing files in plenty of directories.

Comment (jtanx) [reply desktop/Makefile.am]: I went with just dropping them.

**Comment by jtanx** [comment]
As usual, I wasted a bunch of time battling the build system... Anyway, I think it works now. I tested on the `travis` branch, https://travis-ci.org/fontforge/fontforge/jobs/504229500 was green.

AppImages are built only when building from master. They're also pushed to bintray, because transfer.sh is broken (and seems kinda dodgy to rely on).

First working build here: http://dl.bintray.com/fontforge/fontforge/FontForge-c6f12f6-x86_64.AppImage

**Comment by probonopd** [comment]
Where can the AppImages be downloaded from? https://bintray.com/fontforge/fontforge/fontforge-travis does not show them.

**Comment by probonopd** [comment]
> First working build here: http://dl.bintray.com/fontforge/fontforge/FontForge-c6f12f6-x86_64.AppImage

The link does not work.

**Comment by probonopd** [comment]
http://dl.bintray.com/fontforge/fontforge/ has no AppImage.

**Comment by jtanx** [comment]
They'll be there, I deleted all the test builds. I screwed up .travis.yml in another commit, so will have to wait for the next master build to get one.

**Comment by JoesCat** [comment]
putting together a subset of #3387 - basically - just the themes section as we're already gone past the point of fontforge vs org.fontforge.FontForge.

I've cloned appdata into metainfo, this should allow metainfo to progress and if necessary appdata regress.
